### PR TITLE
Update SCP to v16.0.0

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
In case I don't finish tracking down the bugs someone else could take a look at it too.

Related to: https://github.com/bosagora/agora/issues/505

NOTES:

- Targeting v16.0.0 rather than v17.0.0 as the newest one uses C++17 and I'm not sure about the compatibility with D yet.

CHANGES:

- Removed `atomic<bool>` from the quorum intersection checker because it's unnecessary for us. Our quorum generator should not produce giant quorums, unlike in Stellar where they're mostly user-defined (to an extent).

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
The nomination has been fixed. But now `combineCandidates` segfaults. It's probably something to do with the data size & layout of the structs.

**Comment by AndrejMitrovic** [comment]
Found the second bug. I was passing by value instead of by ref. The simple test now passes. I'll let the CI do its run.

**Comment by AndrejMitrovic** [comment]
Passes the CI on ubuntu. I still have to re-enable the quorum checker code, but I think I need https://github.com/bosagora/agora/pull/2066 first to resolve the runtime bugs with using `unorered_map`.

**Comment by AndrejMitrovic** [comment]
I've rebased it against v0.x.x but there's many linker errors. The issue is missing instantiations of `cppSizeOf` and a few other symbols. It's a royal PITA trying to instantiate the right macros.

**Comment by AndrejMitrovic** [comment]
@omerfirmak any idea why I'm still getting this linker error with the last commit?:

>   "int getCPPSizeof<std::__1::unordered_map<int, int, std::__1::RandHasher<int, std::__1::hash<int> >, std::__1::equal_to<int>, std::__1::allocator<std::__1::pair<int const, int> > > >()", referenced from:

File: source/scpd/Cpp.d [hunk 1]
```diff
@@ -89,6 +89,8 @@ extern(C++, (StdNamespace)) extern(C++, class) struct allocator (T) {}
 
 extern(C++, (StdNamespace)) extern(C++, class) struct less (T) {}
 
+extern(C++, (StdNamespace)) extern(C++, class) struct RandHasher (T, Hasher = hash!T) { }
```
Comment (omerfirmak) [inline source/scpd/Cpp.d]:
```suggestion
extern(C++, `std`) extern(C++, class) struct RandHasher (T, Hasher = hash!T) { }
```

should fix it i guess

Comment (AndrejMitrovic) [reply source/scpd/Cpp.d]: Hey that worked! Thanks!

**Comment by Geod24** [comment]
> +3,420 −1,079

Walter, is that you ? :trivial:

**Comment by Geod24** [comment]
For anyone ~mad~ kind enough to review: https://github.com/stellar/stellar-core/compare/v11.2.0...v16.0.0

File: source/scpd/scp/SCPDriver.d [hunk 2]
```diff
@@ -108,6 +155,12 @@ nothrow:
         void* toShortString(ref const(PublicKey) pk) const;
     }
 
+    // `getHashOf` computes the hash for the given vector of byte vector
+    abstract Hash getHashOf(ref vector!Value vals) const;
+
+    // Agora: routing through xdr_to_opaque to get the same hashing behavior
+    Hash getHashOfQuorum(ref const(SCPQuorumSet) qSet) const;
```
Comment (Geod24) [inline source/scpd/scp/SCPDriver.d]: Move the `@trusted` here ?

File: source/scpp/src/util/Logging.cpp [hunk 3]
```diff
@@ -19,8 +19,8 @@ DLogger::DLogger(int level, std::string const& loggerName)
 
 DLogger::~DLogger()
 {
-    if (mLevel == TRACE && !Logging::logTrace("SCP"))
-        return;
+    //if (mLevel == TRACE && !Logging::logTrace("SCP"))
+    //    return;
```
Comment (Geod24) [inline source/scpp/src/util/Logging.cpp]: What's the story here ?

File: source/scpp/src/scp/QuorumSetUtils.cpp [hunk 4]
```diff
@@ -34,47 +35,37 @@ class QuorumSetSanityChecker
...
+            "Total number of nodes in a quorum must be within 1 and 1000";
     }
-    else if (mCount > 1000)
-    {
-        *reason = "Number of validator nodes exceeds the limit of 1000";
-        mIsSane = false;
-    }
-
-    // only one of the two may be true
-    assert(mIsSane ^ (*reason != nullptr));
```
Comment (Geod24) [inline source/scpp/src/scp/QuorumSetUtils.cpp:62]: Such a loss

File: source/scpd/Cpp.d [hunk 5]
```diff
@@ -113,7 +115,7 @@ extern(C++, (StdNamespace)) {
         else
             private alias TPtr = T*;
 
-        mixin CPPBindingMixin!(shared_ptr!T);
+        mixin CPPBindingMixin!(typeof(this));
```
Comment (Geod24) [inline source/scpd/Cpp.d]: You could also use:
```suggestion
        mixin CPPBindingMixin!(shared_ptr);
```

As the template name refers to the current instantiation. I'm curious why the change was necessary though ?

File: source/agora/consensus/protocol/Nominator.d [hunk 6]
```diff
@@ -1231,6 +1241,38 @@ extern(D):
...
+        }
+        void trusted () @trusted
+        {
+            crypto_generichash_final(&state, hash[].ptr, Hash.sizeof);
+        }
+        trusted();
+
+        ubyte[64] bytes = hash[][0..64];
+        return StellarHash(bytes);
+    }
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d:1275]: I think we can improve this but I need to look at it with fresh eyes

File: source/scpd/Cpp.d [hunk 7]
```diff
@@ -151,14 +153,14 @@ private pure nothrow @nogc @safe extern(C++) size_t cpp_unordered_map_length (K,
 
 /// Rudimentary bindings for std::unordered_map
 extern(C++, (StdNamespace))
-public struct unordered_map (Key, T, Hash = hash!Key, KeyEqual = equal_to!Key, Allocator = allocator!(pair!(const Key, T)))
+public struct unordered_map (Key, T, Hash = RandHasher!(Key), KeyEqual = equal_to!Key, Allocator = allocator!(pair!(const Key, T)))
```
Comment (Geod24) [inline source/scpd/Cpp.d:156]: Did we get this wrong or it's a new thing in C++ ?

File: source/scpd/Cpp.d [hunk 8]
```diff
@@ -532,6 +534,13 @@ unittest
 /// Invoke an std::function pointer (note: must be void* due to mangling issues)
 extern(C++) void callCPPDelegate (void* cb);
 
+public mixin template NonMovableOrCopyable ()
+{
+    @disable this ();
+    @disable this (this);
+    @disable ref typeof (this) opAssign () (auto ref typeof(this) rhs);
+}
```
Comment (Geod24) [inline source/scpd/Cpp.d:542]: 👍

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 9]
```diff
@@ -27,7 +27,8 @@ public abstract class QuorumIntersectionChecker
   public:
     /// Create & initialize a QuorumIntersectionChecker with the given map
     static shared_ptr!QuorumIntersectionChecker create (
-        ref const(QuorumTracker.QuorumMap) map);
+        ref const(QuorumTracker.QuorumMap) map,
+	bool quiet = false);
```
Comment (Geod24) [inline source/scpd/quorum/QuorumIntersectionChecker.d:31]: Indent

File: source/scpd/quorum/QuorumTracker.d [hunk 10]
```diff
@@ -69,6 +84,14 @@ public:
...
     /// returns the known quorum map of the entire network
     ref const(QuorumMap) getQuorum () const;
+
+    // Given a node in the transitive quorum, but _outside_ of the local quorum
+    // set, find all nodes in the local quorum set with the shortest distance to
+    // `nodeOutsideQset`. If a node from the local quorum set is passed in,
+    // return itself. If the local node is passed in, return an empty set.
+    // todo:
+    //std::set<NodeID> const&
+    //findClosestValidators(NodeID const& nodeOutsideQset);
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d]: Not needed ?

File: source/scpd/scp/BallotProtocol.d [hunk 11]
```diff
@@ -57,22 +57,39 @@ extern(C++, class) public struct BallotProtocol
...
+    extern (C++, class) public struct SCPBallotWrapper
+    {
+    extern(C++):
+        ValueWrapperPtr mWvalue;
+        const SCPBallot ballot;
+
+      public:
+        this(uint c, ValueWrapperPtr vw);
+        this(ref const(SCPBallotWrapper) o);
+        ~this();
```
Comment (Geod24) [inline source/scpd/scp/BallotProtocol.d]: Don't see the dtor on the C++ side, I assume it's implicit due to fields ?

File: source/scpd/scp/BallotProtocol.d [hunk 12]
```diff
@@ -297,6 +320,12 @@ extern(C++, class) public struct BallotProtocol
     void startBallotProtocolTimer();
     void stopBallotProtocolTimer();
     void checkHeardFromQuorum();
+
+    SCPBallotWrapperUPtr makeBallot(SCPBallot* b) const;
+    SCPBallotWrapperUPtr makeBallot(uint c, Value* v) const;
+
+    // todo
+    //std::string ballotToStr(SCPBallotWrapperUPtr const& ballot) const;
```
Comment (Geod24) [inline source/scpd/scp/BallotProtocol.d]: What's the issue here ?

File: source/scpd/scp/SCPDriver.d [hunk 13]
```diff
@@ -23,6 +23,47 @@ import core.stdc.inttypes;
...
+    ref const(SCPEnvelope) getEnvelope() const;
+    ref const(SCPStatement) getStatement() const;
+}
+
+alias SCPEnvelopeWrapperPtr = shared_ptr!SCPEnvelopeWrapper;
+
+extern (C++, class) public struct WrappedValuePtrComparator
+{
+extern(C++):
+    bool operator()(ref const(ValueWrapperPtr) l, ref const(ValueWrapperPtr) r) const;
```
Comment (Geod24) [inline source/scpd/scp/SCPDriver.d]: Should be `opCall`

File: source/scpd/tests/GlueTypes.d [hunk 14]
```diff
@@ -59,6 +65,20 @@ alias TypesNoLayout = AliasSeq!
...
+    //ValueWrapperPtr,
+    //BallotProtocol.SCPBallotWrapper,
+    //SCPEnvelopeWrapper,
+    //ValueWrapper,
+
+    //LocalNode,
+    //BallotProtocol,
+    //NominationProtocol,
+    //SCP,
+    //Slot,
```
Comment (Geod24) [inline source/scpd/tests/GlueTypes.d]: Size is off ?

**Review by Geod24 — COMMENTED** [review]
Did a first pass. I'm wondering, isn't the only reason we're adding SHA / Curve25519 because of the hashing behavior of Stellar ? IIRC the changes to get rid of those are minimal, as we don't need to match the hashes, just to provide unique hashes (and it's easier nowadays as we have the same hash size).

File: source/scpd/Cpp.d (hunk 5 above)
Comment (AndrejMitrovic) [reply source/scpd/Cpp.d]: Ah some of the instantiations only partially passed the parameters (not all of the template parameters). I'll use the template name trick though.

File: source/scpp/src/util/Logging.cpp (hunk 3 above)
Comment (AndrejMitrovic) [reply source/scpp/src/util/Logging.cpp]: Debugging, will revert.

**Comment by AndrejMitrovic** [comment]
> Did a first pass. I'm wondering, isn't the only reason we're adding SHA / Curve25519 because of the hashing behavior of Stellar ?

It seems there were some pointless imports to SHA.h in most of the modules. But if I remove them it still compiles. So I'll get rid of these modules.

File: source/agora/consensus/protocol/Nominator.d (hunk 6 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:1275]: It would be nice if the hashing routines had support for ranges. I want to be able to do something like this:

```D
const hash = hashFull(vals[].map!(val => val[]));
```

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:1275]: I wonder if I just have to flatten the range.. 🤔

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:1275]: There is a very strange error when passing ranges to `hashFull`:

```
submodules/crypto/source/agora/crypto/Hash.d(177,22): Error: expression `*record` is `void` and has no value
```

But this part of code is:

```D
// Pointers are handled as arrays, only their size must be 0 or 1
else static if (is(T : E*, E))
{
    if (record is null)
        hashVarInt(uint(0), hasher);
    else
    {
        hashVarInt(uint(1), hasher);
        hashPart(*record, hasher);
    }
}
```

How did a range got converted into a pointer? Could it be something to do with `in`?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:1275]: Anyway I think we can add `isInputRange` as a special case in `hashPart` and it could work?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:1275]: Hmm not really possible with input ranges because we only know the length after exhaustion of the input range, but the hash routine should hash the length first.

File: source/scpd/tests/GlueTypes.d (hunk 14 above)
Comment (AndrejMitrovic) [reply source/scpd/tests/GlueTypes.d]: I've re-enabled most of them except two, which I haven't been able to match the size of on mac/linux reliably without guessing the size. Maybe @omerfirmak wants to take a look at these later.

**Comment by AndrejMitrovic** [comment]
Most comments resolved!

File: source/scpd/Cpp.d (hunk 7 above)
Comment (Geod24) [reply source/scpd/Cpp.d:156]: Ping

File: source/scpd/quorum/QuorumIntersectionChecker.d (hunk 9 above)
Comment (Geod24) [reply source/scpd/quorum/QuorumIntersectionChecker.d:31]: Ping

File: source/scpd/Cpp.d (hunk 7 above)
Comment (AndrejMitrovic) [reply source/scpd/Cpp.d:156]: It's a new thing in SCP:

https://github.com/stellar/stellar-core/blob/master/src/util/UnorderedSet.h
https://github.com/stellar/stellar-core/blob/master/src/util/UnorderedMap.h

I haven't looked into why they switched things.

File: source/scpp/src/crypto/XDRHasher.h [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+#include <xdrpp/endian.h>
+#include <xdrpp/marshal.h>
+
+namespace stellar
+{
+
+// CRTP base class to use in XDR-hashing xdrpp-compatible archiver objects. All
+// you need to provide is a `void Derived::hashBytes(unsigned char const*,
+// size_t)` member.
+template <typename Derived> struct XDRHasher
```
Comment (Geod24) [inline source/scpp/src/crypto/XDRHasher.h:16]: Do we need this (and shorthash / siphash) ?

Comment (AndrejMitrovic) [reply source/scpp/src/crypto/XDRHasher.h:16]: I think the problematic part is the QuorumIntersectionChecker. It's technically part of Herder and not SCP, so it brings in a lot of dependencies. For example:

ShortHash uses XdrHasher & SipHash
HashOfHash uses ShortHash
SecretKey uses HashOfHash
QuorumIntersectionChecker uses SecretKey

File: source/scpp/src/scp/LocalNode.h [hunk 16]
```diff
@@ -9,6 +9,7 @@
 #include <vector>
 
 #include "scp/SCP.h"
+#include "util/HashOfHash.h"
```
Comment (Geod24) [inline source/scpp/src/scp/LocalNode.h:12]: Okay so I was looking at this locally. This is what pulls in the hash & secretkey dependency AFAICS.
If you look at our fork of SCP, I actually removed it. I believe it could be removed upstream, but it is used in the implementation (the `.cpp`). 

However, I think those are just lingering import. If you remember, Stellar reduced dependencies by exposing a `sign` method, instead of relying on `SecretKey`. So I believe we should be able to remove all `SecretKey`-related code from our code, but we'll have to remove some (unused) imports.

**Review by Geod24 — APPROVED** [review]
Amazing work, thanks! Let's get it in, and remove the unused dependencies later...

**Comment by Geod24** [comment]
Note that you didn't edit `dub.json` so all the new `.cpp` files are actually not linked in.

File: source/scpp/src/util/Math.cpp [hunk 17]
```diff
@@ -17,20 +22,135 @@ rand_fraction()
...
+        return std::set<double>(points.begin(), points.end());
+    }
+
+    std::set<double> centroids;
+
+    auto backlog = points;
+
+    auto moveIndexToCentroid = [&](size_t index) {
+        // @BUG: undefined symbol printAssertFailureAndThrow()
+        // releaseAssertOrThrow(index < backlog.size());
```
Comment (Geod24) [inline source/scpp/src/util/Math.cpp:72]: This can be uncommented if we add `GlobalChecks.o` to `dub.json`.
But I'd like to either get rid of it (and the `Backtace.cpp` dependency) or have it implemented in D.
